chain/network/src/peer/stream.rs

```diff
@@ -0,0 +1,203 @@
+    let buf_size_metric = metrics::MetricGuard::new(
+        &metrics::PEER_DATA_READ_BUFFER_SIZE,
+        vec![peer_addr.to_string()],
+    );
+    loop {
+        let n = read.read_u32_le().await.map_err(RecvError::IO)? as usize;
+        if n > NETWORK_MESSAGE_MAX_SIZE_BYTES {
+            return Err(RecvError::MessageTooLarge {
+                got_bytes: n,
+                want_max_bytes: NETWORK_MESSAGE_MAX_SIZE_BYTES,
+            });
+        }
+        msg_size_metric.observe(n as f64);
+        buf_size_metric.set(n as i64);
+        let mut buf = vec![0; n];
+        let t = metrics::PEER_MSG_READ_LATENCY.start_timer();
+        read.read_exact(&mut buf[..]).await.map_err(RecvError::IO)?;
+        t.observe_duration();
+        buf_size_metric.set(0);
+        stats.received_messages.fetch_add(1, Ordering::Relaxed);
+        stats.received_bytes.fetch_add(n as u64, Ordering::Relaxed);
+        if let Err(_) = addr.send(Frame(buf)).await {
+            // We got mailbox error, which means that Actor has stopped,
+            // so we should just close the stream.
+            return Ok(());
+        }
+    }
+}
+async fn run_send_loop(
+    tcp_send: WriteHalf,
+    mut queue_recv: tokio::sync::mpsc::UnboundedReceiver<Vec<u8>>,
+    stats: Arc<connection::Stats>,
+    buf_size_metric: Arc<metrics::IntGaugeGuard>,
+) -> io::Result<()> {
+    const WRITE_BUFFER_CAPACITY: usize = 8 * 1024;
+    let mut writer = tokio::io::BufWriter::with_capacity(WRITE_BUFFER_CAPACITY, tcp_send);
+    loop {
+        tokio::select! {
+            biased;
+            // Receive the next message.
+            // "biased" should ensure that write.flush() won't be polled at all
+            // if there is a ready message in the channel.
+            msg = queue_recv.recv() => { match msg {
+                Some(msg) => {
+                    // Write the frame to the buffer.
+                    writer.write_u32_le(msg.len() as u32).await?;
+                    writer.write_all(&msg[..]).await?;
+                    stats.messages_to_send.fetch_sub(1,Ordering::Release);
+                    stats.bytes_to_send.fetch_sub(msg.len() as u64,Ordering::Release);
+                    buf_size_metric.sub(msg.len() as i64);
+                }
+                // There will be no new messages: flush and return.
+                None => return writer.flush().await,
+            } }
+            // Flush the buffer iff it is non-empty.
+            res = writer.flush(), if writer.buffer().len()>0 => { res?; }
+        }
+    }
+}
+
+// Stream critical error.
+// PeerActor is responsible for calling ctx.stop() after receiving stream::Error.
+#[derive(thiserror::Error, Debug, actix::Message)]
+#[rtype(result = "()")]
+pub(crate) enum Error {
+    #[error("send")]
+    Send(#[source] SendError),
+    #[error("recv")]
+    Recv(#[source] RecvError),
+}
+
+pub(crate) struct FramedStream {
```

Comment:
Continuing spawn/run naming conventions, I like to call things which "own" concurrent resources `Handle`. Ie, the actual stream are the two futures we spawn, this is just a handle to communicate with those.